### DISCUSSION TITLE
[Design] Implant Config Management

### DISCUSSION THREAD

**Discussion by KCarretto** [spellshift/realm#24] [Ideas]
How should Implant Configs be created and managed?

**Goals**
- Track the status of implants across targets (even those that do not callback)
- Update configuration information for implants during an engagement
- Visualize current-state of configuration information in the Tavern UI

**Comment by KCarretto** [comment]
From #18:

> I STRONGLY advise against the language setting state of the bot. If it would update a config file that’s different but updating the bot’s internal variables is just a mess going forward. We could also manage these things via the config.yml

We'll need a method to update a running agent's configuration, e.g. it's callback interval. If we have this API update a config file for the implant, is the implant watching for updates to the config file?

**Reply by hulto** [reply]
I think having a built-in function to "reconfigure" that re-reads the config file could be a straight forward approach to this.
Not a priority now but something we may wish to do is make an option to store the config ephemerally in memory or encrypt it on disk. I can see blue teams greping / for specific strings to identify the imix agent.

**Comment by hulto** [comment]
@KCarretto RE: Implant registration.

> This warrants some further discussion, but I imagine some type of implant standard library, which could include functions for registration and config management. For example:
> 
> implant.register("imix")
> implant.set_interval(120)
> implant.add_transport("realm-c2", "https://c2.realm.pub", proxy="https://my.proxy.org", priority=5)
> implant.remove_transport("realm-c2")
> file.write("/etc/imix.conf", implant.get_config())


Instead of creating a registration spell can we simplify and have that be the same as reconfiguration?
Is there anything specific to a registration action that we wouldn't want to be updated during reconfig?
